File: scripts/data/lib/webp.py
"""Encode one image to WebP under a hard byte cap and a hard pixel cap.

Called by `lib/webp.mjs`. Kept in Python because Phase 8 encoded all 34 pilot
images with Pillow and the alternative is a ~30 MB native npm dependency for a
step that only ever runs at authoring time — `scripts/data/` is never run by CI
or by the build (Instruction.md Phase 9, SPEC.md §13).

The rule this file exists to enforce is FRICTION-LOG D3: **never pick a fixed
quality.** The cap is on bytes, so a busy photograph at a fixed quality misses
it while a plain one wastes half the budget. Quality is searched downward, and
only when quality alone cannot reach the cap is the image resized — in that
order, because losing pixels is worse than losing a little chroma detail on a
photograph of a machined object.

Prints one JSON object to stdout. Every diagnostic goes to stderr, so the
caller can parse stdout unconditionally.

    python webp.py <input> <output> --max-px 1600 --max-bytes 204800
"""

import argparse
import io
import json
import sys

try:
    from PIL import Image, ImageOps
except ImportError:  # pragma: no cover - environment problem, not a code path
    print(
        "Pillow is not installed. `pip install Pillow` — the image pipeline "
        "needs it to produce WebP (SPEC.md §10).",
        file=sys.stderr,
    )
    raise SystemExit(2)

# Searched in this order. 86 first because it cleared the gallery cap for 30 of
# the 34 pilot images; the tail exists because the other four needed 78 and 45.
QUALITIES = (92, 86, 80, 74, 68, 62, 56, 50, 45, 40, 35, 30)

# Applied only after every quality has failed. Each step is a fresh resize from
# the ORIGINAL, never a resize of an already-resized copy: repeated resampling
# softens an image visibly and there is no reason to accept that here.
SCALE_STEPS = (1.0, 0.85, 0.72, 0.6, 0.5, 0.42, 0.35)


def encode(image, quality):
    buffer = io.BytesIO()
    # method=6 is the slowest and smallest setting. At one image per invocation
    # the extra second is free and it is worth 5-10% of the byte budget.
    image.save(buffer, format="WEBP", quality=quality, method=6)
    return buffer.getvalue()


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("source")
    parser.add_argument("destination")
    parser.add_argument("--max-px", type=int, required=True)
    parser.add_argument("--max-bytes", type=int, required=True)
    args = parser.parse_args()

    with Image.open(args.source) as raw:
        # EXIF orientation is applied here rather than trusted: a phone-camera
        # upload on Commons will otherwise be stored rotated, and the width and
        # height written into the imageRef would describe the wrong axes.
        original = ImageOps.exif_transpose(raw)
        original = original.convert("RGBA" if "A" in original.getbands() else "RGB")

        longest = max(original.size)
        base_scale = min(1.0, args.max_px / longest)

        for step in SCALE_STEPS:
            scale = base_scale * step
            width = max(1, round(original.width * scale))
            height = max(1, round(original.height * scale))
            resized = (
                original
                if (width, height) == original.size
                else original.resize((width, height), Image.LANCZOS)
            )
            for quality in QUALITIES:
                data = encode(resized, quality)
                if len(data) <= args.max_bytes:
                    with open(args.destination, "wb") as out:
                        out.write(data)
                    print(
                        json.dumps(
                            {
                                "width": width,
                                "height": height,
                                "bytes": len(data),
                                "quality": quality,
                                "sourceWidth": original.width,
                                "sourceHeight": original.height,
                                "resized": (width, height) != original.size,
                            }
                        )
                    )
                    return

    print(
        f"could not reach {args.max_bytes} bytes for {args.source} even at "
        f"quality {QUALITIES[-1]} and {SCALE_STEPS[-1]:.0%} scale. This is almost "
        "always a photograph with heavy noise or a busy background — pick a "
        "different file rather than degrading this one further.",
        file=sys.stderr,
    )
    raise SystemExit(1)


if __name__ == "__main__":
    main()

File: scripts/data/lib/webp.mjs
/**
 * WebP encoding, with the caps read from the same file `check:assets` reads.
 *
 * `src/data/thresholds.json` is the single source of the byte and pixel caps,
 * and of the "a file called `hero.webp` is the hero" rule. The encoder and the
 * gate must not each carry their own copy: an encoder targeting 200 KB while
 * the gate fails at 120 KB produces files that pass locally and break the
 * build, which is precisely the drift Phase 2 wrote that file to prevent.
 *
 * Encoding itself is `webp.py` — see its header for why it is Python.
 */
import { spawn } from 'node:child_process';
import { readFile, stat, writeFile, mkdir, unlink } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

export const REPO_ROOT = path.resolve(fileURLToPath(new URL('../../..', import.meta.url)));

const thresholds = JSON.parse(
  await readFile(path.join(REPO_ROOT, 'src', 'data', 'thresholds.json'), 'utf8'),
).assets;

export const ASSET_CAPS = thresholds;

/**
 * Which caps apply to a file, decided by its basename exactly as
 * `check:assets` decides it.
 */
export function capsForBasename(basename) {
  const isHero = basename.toLowerCase() === `hero.${thresholds.format}`;
  return isHero
    ? { role: 'hero', maxBytes: thresholds.heroMaxBytes, maxPixels: thresholds.heroMaxPixels }
    : {
        role: 'gallery',
        maxBytes: thresholds.galleryMaxBytes,
        maxPixels: thresholds.galleryMaxPixels,
      };
}

function runPython(args, stdin = null) {
  return new Promise((resolve, reject) => {
    // `python` rather than `python3`: this is a Windows-first repo
    // (Instruction.md §1) and `python3` is not on PATH there.
    const child = spawn('python', args, {
      // Console code page is 1256 on this machine and a Commons filename will
      // contain characters it cannot represent (Instruction.md §1).
      env: { ...process.env, PYTHONIOENCODING: 'utf-8' },
    });
    if (stdin !== null) {
      child.stdin.on('error', () => {}); // a python that died early closes the pipe
      child.stdin.end(stdin, 'utf8');
    }
    let out = '';
    let err = '';
    child.stdout.on('data', (d) => (out += d));
    child.stderr.on('data', (d) => (err += d));
    child.on('error', (error) =>
      reject(
        new Error(
          `could not run python (${error.message}). The image pipeline needs Python with Pillow; ` +
            'it is authoring-time only and no part of the build or CI depends on it.',
        ),
      ),
    );
    child.on('close', (code) => {
      if (code === 0) resolve(out.trim());
      else reject(new Error(err.trim() || `python exited ${code}`));
    });
  });
}

/**
 * Tiles candidate images into one numbered contact sheet, for triage.
 *
 * `PLAYBOOK.md` §6 requires three searches per subject, so the agent routinely
 * has more candidates than it can afford to open. This writes one small sheet
 * instead; only the finalist is fetched at full size. Triage only — the tiles
 * cannot show a watermark or which variant a receiver is.
 *
 * Takes `[{ label, bytes }]` and returns what the tiler actually drew, so a
 * candidate that will not decode is reported rather than silently missing.
 */
export async function writeSheet(tiles, destination) {
  await mkdir(path.dirname(destination), { recursive: true });
  const job = JSON.stringify({
    out: destination,
    tiles: tiles.map(({ label, bytes }) => ({
      label,
      bytes_b64: Buffer.from(bytes).toString('base64'),
    })),
  });
  const script = path.join(REPO_ROOT, 'scripts', 'data', 'lib', 'sheet.py');
  const out = await runPython([script], job);
  return JSON.parse(out);
}

/**
 * Encodes `bytes` to WebP at `destination`, under that filename's caps.
 *
 * Returns the **measured** width, height and byte size of the file actually
 * written. Never the requested ones: Markey's whole first batch of image
 * dimensions was wrong because they were computed from the request rather than
 * read back from the file, and a `width` nobody measured is not a measurement
 * (the same sentence is in `check:assets`).
 */
export async function encodeToWebp(bytes, destination, { overwrite = false } = {}) {
  const caps = capsForBasename(path.basename(destination));
  await mkdir(path.dirname(destination), { recursive: true });

  const existing = await stat(destination).catch(() => null);
  if (existing && !overwrite) {
    throw new Error(
      `${destination} already exists (${Math.round(existing.size / 1024)} kB). ` +
        'The last argument is a filename, not a folder — use hero, or a descriptive name per image. ' +
        'SPEC.md §10: never re-encode in place; a changed image is a new filename.',
    );
  }

  const temp = `${destination}.source.tmp`;
  await writeFile(temp, bytes);
  try {
    const output = await runPython([
      path.join(REPO_ROOT, 'scripts', 'data', 'lib', 'webp.py'),
      temp,
      destination,
      '--max-px',
      String(caps.maxPixels),
      '--max-bytes',
      String(caps.maxBytes),
    ]);
    const result = JSON.parse(output);
    const measured = (await stat(destination)).size;
    if (measured > caps.maxBytes) {
      throw new Error(`encoded ${measured} bytes, over the ${caps.role} cap of ${caps.maxBytes}`);
    }
    return { ...result, bytes: measured, role: caps.role };
  } finally {
    await unlink(temp).catch(() => {});
  }
}
